android/jni/app-android.cpp

```diff
@@ -612,16 +620,23 @@ extern "C" void Java_org_ppsspp_ppsspp_NativeRenderer_displayInit(JNIEnv * env,
 	if (javaGL && !graphicsContext) {
 		graphicsContext = new AndroidJavaEGLGraphicsContext();
 	}
-	ILOG("NativeApp.displayInit() (graphicsContext=%p)", graphicsContext);
 
-	if (!renderer_inited) {
+	if (renderer_inited) {
+		ILOG("NativeApp.displayInit() restoring");
+		NativeDeviceLost();
+		NativeShutdownGraphics();
+		NativeDeviceRestore();
+		NativeInitGraphics(graphicsContext);
```

Comment:
It's a little tricky, we need to do the restore first, otherwise the "preset" shaders of thin3d will not be restored, which will cause things to crash because the programs will fail to link when NativeInitGraphics creates pipelines. I don't think anything can access g_gameInfoCache during the restore.